<a href="https://colab.research.google.com/github/R0N4L2/alaYa/blob/main/2PreguntasdeModelamientoyProgramacion_RonaldCastillo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

__Pregunta1:__

Una línea áerea decide incorporar un nuevo avión para vuelos comerciales a su flota. Para habilitarlo necesita conocer cuáles el número óptimo de asientos por clase $(x_i)$, la cantidad  requerida de azafatas $(y_1)$ y auxiliares de vuelo $(y_2)$, de tal manera que la utilidad sea máxima. Existen 3 tipos de clases: Primera, Ejecutiva y Económica. Por políticas internas se debe ofrecer un mínimo de asientos por clase de 25, 80 y 120 respectivamente. Además un estudio de mercado indicó que la demanda máxima para cada clase es de 45, 100 y 210, por lo que tener un número superior de asientos por clase no tendrá ningún sentido. Por otra parte, el número de azafatas y auxiliares también está acotado. Por un lado, el avión no puede funcionar con menos de 8 azafatas y 2 auxiliares de vuelo; y por límite de espacio no podrán ser más de 18 azafatas y 5 auxiliares de vuelo. Además, para entregar un buen servicio en cada clase, deberá haber almenos 1 azafata por cada 10 pasajeros de Primera, por cada 20 de Ejecutiva y por cada 40 de Económica. También deberá haber un auxiliar de vuelo por cada 100 pasajeros del avión. El sueldo de cada azafata es de $\$200$ dólares y el de un auxiliar de $\$120$ dólares. El avión dispone de $420 m^2$ para distribuir los asientos (el espacio para pasillos, cabinas y baños no está incluido), y habrá que considerar que un  asiento de Primera ocupa $1.8 m^2$, uno de Ejecutiva $1.4m^2$ y unode Económica $1 m^2$. El valor de  un pasaje en cada una de las claseses de $\$2,000$, $\$1,300$ y $\$900$ dólares respectivamente, mientras  que el costo de la comida para cada una de las clases es de $\$80$, $\$60$ y $\$50$ dólares. El costo de mantención del avión es de $\$75,000$ dólares. Finalmente, tras un cuidadoso estudio de servicio e imagen, la línea aérea concluyó que por cada azafata que tuviera por sobre el mínimo, recibiría un beneficio total equivalente a $\$100$ dólares, y por cada auxiliar de vuelo adicional el beneficio sería de $\$50$ dólares; esto debido a que entregarían un mejor servicio y los clientes preferirían viajar en su línea. Se pide:

* Modelar el problema de manera extendida. 
* Programar y resolver mediante un optimizador de Python (Pulp,Pyomo,PythonGurobio similares).

Se debe indicar el valor de cada variable y su función objetivo.

__Respuesta__:

El problema de optimización lineal puede ser formulado de la siguiente manera:

* _Variables de decisión_:

  1. $x_i$: Número de asientos por clase (Primera[i=1], Ejecutiva [i=2] y Económica [i=3]).
  2. $y_1$: Cantidad de azafatas.
  3. $y_2$: Cantidad de auxiliares de vuelo.

* _Función objetivo_:

  Maximizar la utilidad total de la línea aérea, considerando los ingresos por pasajes, el costo de mantención del avión y los beneficios adicionales por tener azafatas y auxiliares de vuelo extra. La función objetivo se puede expresar como:

  ___Maximizar___:

$$f(x,y)=(2,000-80) * x_1 + (1,300-60) * x_2 + (900-50) * x_3 - 75,000 - 300 * y_1 - 170 * y_2$$

*_Restricciones_:

 Límite de asientos por clase:

 1. $45 \geq x_1 \geq 25$ (Primera)
 2. $100 \geq x_2 \geq 80$ (Ejecutiva)
 3. $210 \geq x_3 \geq 120$ (Económica)
 4. $18 \geq y_1 \geq 8$ (azafatas)
 5. $5 \geq y_2 \geq 2$ (auxiliares de vuelo)
 
  Restricciones de personal por número de pasajeros:
  6. $y_1 \geq \frac{x_1}{10}+\frac{x_2}{20}+\frac{x_2}{40}$
  
  Restricción de espacio para asientos:
  7. $1.8 * x_1 + 1.4 * x_2 + 1 * x_3 = 420$

  * Notas adicionales: Las variables $x_1$, $x_2$, $x_3$, $y_1$ y $y_2$ deben ser números enteros.

In [17]:
from scipy.optimize import linprog
c = [-1920,-1240,-850,300,170]
A = [[1, .5,.25,-10,0]]
b = [0]
Ae=[[1.8,1.4,1,0,0]]
be=[420]
x1_bounds = (25,45)
x2_bounds = (80,100)
x3_bounds = (120,210)
y1_bounds = (8,18)
y2_bounds = (2,5)
res = linprog(c, A_ub=A, b_ub=b,A_eq=Ae, b_eq=be,bounds=[x1_bounds, x2_bounds,x3_bounds,y1_bounds,y2_bounds],integrality=1)
print(res.x)
print(-res.fun-75000)

[ 45. 100. 199.  15.   2.]
299710.0


Entonces los pasajeros de primera clase son 45, los de clase ejecutiva son 100, los de economica son 199, son 15 azafatas y 2 auxiliares de vuelo, con una ganancia de $\$299710$ dollares.

__Pregunta2__: A continuación, se presenta una escena de un juego de acertijos, donde el jugador requiere seleccionar una combinación correcta de los alimentos para seguir avanzando en su aventura:

El objetivo del puzzle es el siguiente: Se cuenta con 5 tipos distintos de comida (ensalada,sopa, carne,pescado y pasta). Cada uno de ellos tiene un aporte nutricional (naranjo, rojo y verde), el cual se muestra en la siguiente figura:

| plato | aporte | tipo |
|-------|--------|------|
| carne | 3 | verde |
| carne | 3 | rojo |
| carne | 1 | naranja |
| pescado | 2 | verde |
| pescado | 3 | rojo |
| pescado | 1 | naranja |
| ensalada | 0 | verde |
| ensalada | 1 | rojo |
| ensalada | 2 | naranja |
| pasta | 2 | verde |
| pasta | 1 | rojo |
| pasta | 3 | naranja |
| sopa | 1 | verde |
| sopa | 2 | rojo |
| sopa | 1 | naranja |


Debido a esto, el juego requiere que se planifique la cantidad de platos de comida para los 5 días, considerando que en cada día se tiene que alcanzar un requerimiento nutricional distinto, el cualse muestra en la siguiente figura:

| dia (week) | aporte | tipo |
|-------|--------|------|
| A | 8 | verde |
| A | 10 | rojo |
| A | 8 | naranja |
| B | 4 | verde |
| B | 9 | rojo |
| B | 7 | naranja |
| C | 9 | verde |
| C | 10 | rojo |
| C | 8 | naranja |
| D | 9 | verde |
| D | 11 | rojo |
| D | 8 | naranja |
| E | 10 | verde |
| E | 10 | rojo |
| E | 9 | naranja |


Por último, considerar que sólo se cuenta con 5 platos de cada comida. En base a esto, formule un modelo matemático que encuentre una solución.

__Respuesta__:

El problema de optimización lineal puede ser formulado de la siguiente manera:

* _Variables de decisión_:

  1. $x_i$: Número de platos ($1 \leq i \leq 5$).

* _Función objetivo_:

  Maximizar la ganancia nutricional:

  ___Maximizar___:

$$f(x)=\sum  x_i$$

*_Restricciones_:

 Límite de asientos por clase:

 1. $\sum_{k=1}^{5} x_{i,j,k} =5$ 
 2. $\sum_{i=1}^{5}  w_{i} x_{i,j,k} \geq A_{j,k}$ aporte nutricional j para el dia k 
 3. $5 \geq x_{i,k,j}$

  * Notas adicionales: Las variables $x_i$ deben ser números enteros.

In [55]:
from scipy.optimize import linprog
import numpy as np
c = -np.ones(25)
A = np.array([[3,2,0,2,1],[3,3,1,1,2],[1,1,2,3,1]])
Au=-np.kron(np.eye(5),A)
Ae=np.kron(np.ones(5),np.eye(5))
bu = [-8,-10,-8,-4,-9,-7,-9,-10,-8,-9,-11,-8,-10,-10,-9]
be=[5]*5
x_bounds = (0,5)
res = linprog(c, A_ub=Au, b_ub=bu,A_eq=Ae, b_eq=be,bounds=[x_bounds]*25,integrality=1)
print(res.x)
print(-res.fun)

[1. 1. 1. 1. 1. 0. 1. 2. 0. 2. 2. 0. 1. 1. 1. 1. 2. 1. 1. 0. 1. 1. 0. 2.
 1.]
25.0


A continuacion se puede ver en la tabla, como se distribuyen los diferentes platos durante la semana.

In [62]:
import pandas as pd
pd.DataFrame(res.x.reshape(5,5),columns=['A','B','C','D','E'],index=['carne','pescado','ensalada','pasta','sopa']).astype(int)

A  B  C  D  E
carne     1  1  1  1  1
pescado   0  1  2  0  2
ensalada  2  0  1  1  1
pasta     1  2  1  1  0
sopa      1  1  0  2  1